internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,162 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/keys"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+var _ storage.OpenFGADatastore = (*cachedDatastore)(nil)
+
+type cachedDatastore struct {
+	storage.OpenFGADatastore
+
+	cache        storage.InMemoryCache[any]
+	maxCacheSize int64
+	ttl          time.Duration
+}
+
+// NewCachedDatastore returns a wrapper over a datastore...
+func NewCachedDatastore(inner storage.OpenFGADatastore, cache storage.InMemoryCache[any], maxSize int64, ttl time.Duration) *cachedDatastore {
```

Comment:
Please re-use the existing https://github.com/openfga/openfga/blob/5bc7ebb087e56641fccb7027845d3141feacbcff/pkg/storage/storagewrappers/caching.go#L18-L27

Reply:
I wanted to explicitly target storage usage from Check only. For example, Expand API uses `Read()`. Will think a bit to see if I can come up with an alternative.

Reply:
Ohhh I see. Yeah then maybe we can keep them separate.

Reply:
This is something more like `WrapDatastore()`.

Reply:
It just occurred to me: do we have data to prove that calls to Expand are not immediately followed by Checks, or vice-versa? Because if so, it would actually be beneficial for all Check/ListUsers/ListObjects/Expand/Read calls to use the same cache.

Reply:
@miparnisari it could definitely be yeah, but want to keep _this_ change strictly scoped to Check.